metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -1231,3 +1212,87 @@ func AddClusterECSMeta(base mb.BaseMetricSet) mapstr.M {
+	}
+	old.TypeMeta = metav1.TypeMeta{
+		APIVersion: "apps/v1",
+		Kind:       "ReplicaSet",
+	}
+	return old, nil
+}
+
+// getEventMetadataFunc returns a function that takes a kubernetes Resource as an argument and returns metadata
+// that can directly be used for event enrichment.
+// This function is intended to be used as the resource watchers add and update handler.
+func getEventMetadataFunc(
+	logger *logp.Logger,
+	generalMetaGen *metadata.Resource,
+	specificMetaGen metadata.MetaGen,
+	metricsRepo *MetricsRepo,
+) func(r kubernetes.Resource) map[string]mapstr.M {
+	return func(r kubernetes.Resource) map[string]mapstr.M {
+		accessor, accErr := meta.Accessor(r)
+		if accErr != nil {
+			logger.Errorf("Error creating accessor: %s", accErr)
+		}
+		id := accessor.GetName()
+		namespace := accessor.GetNamespace()
+		if namespace != "" {
+			id = join(namespace, id)
+		}
+
+		switch r := r.(type) {
+		case *kubernetes.Pod:
+			return map[string]mapstr.M{id: specificMetaGen.Generate(r)}
+
+		case *kubernetes.Node:
+			nodeName := r.GetObjectMeta().GetName()
+			metrics := NewNodeMetrics()
+			if cpu, ok := r.Status.Capacity["cpu"]; ok {
+				if q, err := resource.ParseQuantity(cpu.String()); err == nil {
+					metrics.CoresAllocatable = NewFloat64Metric(float64(q.MilliValue()) / 1000)
+				}
+			}
+			if memory, ok := r.Status.Capacity["memory"]; ok {
+				if q, err := resource.ParseQuantity(memory.String()); err == nil {
+					metrics.MemoryAllocatable = NewFloat64Metric(float64(q.Value()))
+				}
+			}
+			nodeStore, _ := metricsRepo.AddNodeStore(nodeName)
+			nodeStore.SetNodeMetrics(metrics)
+
+			return map[string]mapstr.M{id: generalMetaGen.Generate(NodeResource, r)}
+		case *kubernetes.Deployment:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(DeploymentResource, r)}
+		case *kubernetes.Job:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(JobResource, r)}
+		case *kubernetes.CronJob:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(CronJobResource, r)}
+		case *kubernetes.Service:
+			return map[string]mapstr.M{id: specificMetaGen.Generate(r)}
+		case *kubernetes.StatefulSet:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(StatefulSetResource, r)}
+		case *kubernetes.Namespace:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(NamespaceResource, r)}
+		case *kubernetes.ReplicaSet:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(ReplicaSetResource, r)}
+		case *kubernetes.DaemonSet:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(DaemonSetResource, r)}
+		case *kubernetes.PersistentVolume:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(PersistentVolumeResource, r)}
+		case *kubernetes.PersistentVolumeClaim:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(PersistentVolumeClaimResource, r)}
+		case *kubernetes.StorageClass:
+			return map[string]mapstr.M{id: generalMetaGen.Generate(StorageClassResource, r)}
+		default:
```

Comment:
Note: We can't have a branch for `PartialObjectMetadata` here, because this may represent different resource types. So this falls into the default branch, and that does the right thing *as long as TypeMeta is set*. This is why our transform function adds the TypeMeta.

Reply:
Does not this fall under the case of kubernetes.replicaset rather than default?

Reply:
No, the actual struct we get here has type `PartialObjectMetadata`.